### PR TITLE
Fees/posting sql error fix

### LINKED ISSUE #871: Fees/Posting input errors
- In Fees/Posting >> submit all empty >> SQL Errors
- In Fees/Posting >> submit with any kind of data >> SQL Errors

**This kind of SQL errors:**

image

### PR DESCRIPTION
**Pull request by MigDinny**
Fixes #871 

_form_portalnotify_ was out of the other _form__ checkings. That means that if _form_portalnotify_ was true, _form_cb_ didn't need to be true to advance. Besides, any _isset()_ wasn't being performed on _form_cb_ and that would cause an SQL error.

This was tested on my computer and no bugs were detected.

**TL;DR**

If no search was performed / checkboxes selected before when clicking on PDF Download, an SQL error would be printed out. This prevents that error: I created a checking when clicking on PDF Download that verifies if any checkbox is selected

### CODE REVIEW

File: interface/billing/sl_eob_search.php [hunk 1]
```diff
@@ -232,8 +232,9 @@ function upload_file_to_client_pdf($file_to_send) {
   
   
     // Print or download statements if requested.
-    //
-  if (($_POST['form_print'] || $_POST['form_download'] || $_POST['form_pdf']) || $_POST['form_portalnotify'] && $_POST['form_cb']) {
+    // This code will only be ran IF form_cb has AT LEAST 1 checkbox selected
+
+    if (($_POST['form_print'] || $_POST['form_download'] || $_POST['form_pdf'] || $_POST['form_portalnotify']) && isset($_POST['form_cb'])) {
```
Comment (teryhill) [inline interface/billing/sl_eob_search.php]: Moving form_portalnotify into the form checking should be the only fix that is required. The isset will take care of a php 7.1 warning.

Comment (MigDinny) [reply interface/billing/sl_eob_search.php]: Should I only remove "isset" ?

Comment (teryhill) [reply interface/billing/sl_eob_search.php]: no leave it .It will have to be addressed by who ever does the fixes for 7.1 compatibility.

Comment (MigDinny) [reply interface/billing/sl_eob_search.php]: No changes necessary then?

**Comment by MigDinny** [comment]
Another bug was fixed as requested by @teryhill

**Comment by MigDinny** [comment]
@teryhill Another bugfix was released

**Comment by teryhill** [comment]
So My fix worked  8^)

**Comment by MigDinny** [comment]
@teryhill yes haha

**Comment by aethelwulffe** [comment]
@MigDinny You have conflicts here now.  Please get them fixed up

File: interface/billing/sl_eob_search.php [hunk 2]
```diff
@@ -1,849 +1,832 @@
...
-        $stmt['amount'] = '0.00';
-        $stmt['ins_paid'] = 0;
-        $stmt['today'] = $today;
-        $stmt['duedate'] = $duedate;
-      } else {
-        // Report the oldest due date.
-        if ($duedate < $stmt['duedate']) {
-          $stmt['duedate'] = $duedate;
+  
+  function bucks($amount) {
```
Comment (aethelwulffe) [inline interface/billing/sl_eob_search.php]: Geez...  function bucks is defined all over the place.  I think there is a better way...one day we will get to that.

File: interface/billing/sl_eob_search.php [hunk 3]
```diff
@@ -1,849 +1,832 @@
...
+    // sleep one second to ensure there's no follow-on.
+    sleep(1);
+  }
+  function upload_file_to_client_pdf($file_to_send) {
+    //Function reads a HTML file and converts to pdf.
+  
+    global $STMT_TEMP_FILE_PDF;
+    global $srcdir;
+    if ($GLOBALS['statement_appearance'] == '1') {
+      require_once("$srcdir/html2pdf/vendor/autoload.php");
```
Comment (aethelwulffe) [inline interface/billing/sl_eob_search.php]: Legacy "plugin" system stuff I suppose.

File: interface/billing/sl_eob_search.php [hunk 4]
```diff
@@ -1,849 +1,832 @@
...
+    $countline=1;
+    $file = fopen($file_to_send, "r");//this file contains the text to be converted to pdf.
+      while(!feof($file)) {
+      $OneLine=fgets($file);//one line is read
+       if(stristr($OneLine, "\014") == true && !feof($file))//form feed means we should start a new page.
+        {
+          $pdf->ezNewPage();
+          $pdf->ezSetY($pdf->ez['pageHeight'] - $pdf->ez['topMargin']);
+          str_replace("\014", "", $OneLine);
         }
```
Comment (aethelwulffe) [inline interface/billing/sl_eob_search.php]: Real changes start here I suppose.

File: interface/billing/sl_eob_search.php [hunk 5]
```diff
@@ -1,849 +1,832 @@
...
-  } // end while
-
-    if (!empty($stmt)) ++$stmt_count;
-    fwrite($fhprint, make_statement($stmt));
-    fclose($fhprint);
+    header("Pragma: public");//this section outputs the pdf file to browser
+    header("Expires: 0");
+    header("Cache-Control: must-revalidate, post-check=0, pre-check=0");
+    header("Content-Type: application/force-download");
+    header("Content-Length: " . filesize($STMT_TEMP_FILE_PDF));
```
Comment (aethelwulffe) [inline interface/billing/sl_eob_search.php]: `header("Content-Length: ".filesize($file));`
We do this forced download PDF thing in a number of places.  It should be turned into a separate library function.  We also still use at least a couple of different PDF output libraries.  No XPS alternative yet (don't really want it anyway) and PostScript is still just Adobe and is too big.  Wish we had another alternative

File: interface/billing/sl_eob_search.php [hunk 6]
```diff
@@ -1,849 +1,832 @@
...
+                if (substr($_POST['form_category'], 0, 3) == 'Due' && !$isdueany) continue;
+                if ($_POST['form_category'] == 'Due Ins' && ($duncount >= 0 || !$isdueany)) continue;
+                if ($_POST['form_category'] == 'Due Pt'  && ($duncount <  0 || !$isdueany)) continue;
+            
+                $bgcolor = ((++$orow & 1) ? "#ffdddd" : "#ddddff");
+            
+                $svcdate = substr($row['date'], 0, 10);
+                $last_stmt_date = empty($row['last_stmt_date']) ? '' : $row['last_stmt_date'];
+            
+                // Determine if customer is in collect
```
Comment (aethelwulffe) [inline interface/billing/sl_eob_search.php]: New line at end of file please.

**Comment by aethelwulffe** [comment]
When you have a lot of whitespace changes, manually adding a reviewer link that ignores whitespace (add "?w=1 " to the URL) makes thing easier for other folks.  I need to remember this cheat myself before I try reviewing something like this.

**Comment by MigDinny** [comment]
@aethelwulffe Didn't understand. Which link?

File: interface/billing/sl_eob_search.php (hunk 4 above)
Comment (MigDinny) [reply interface/billing/sl_eob_search.php]: I can't solve the merge conflict. I don't know what is causing the problem.

Comment (MigDinny) [reply interface/billing/sl_eob_search.php]: I think it's good now.

**Comment by MigDinny** [comment]
I cant understand why this is making a conflict. I've already removed the marks and nothing happens

**Comment by MigDinny** [comment]
@aethelwulffe @teryhill I'm not able to solve this conflict... can any of you give me some help?

**Comment by aethelwulffe** [comment]
@MigDinny Sorry, but I am terrible at resolving conflicts like this.  
https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/
Basically, remove the markers (you can use the web editor here) and change the content manually  to the version you want.
  Just go here:  https://github.com/LibreHealthIO/LibreEHR/pull/885/conflicts
and edit out those markers and change the content.  That should fix it.

**Comment by aethelwulffe** [comment]
Another option is to diff and merge up the two versions of the file into a new file, then add changes across the code base to change the file to another name.  The file name sl_eob_search.php comes from SQL Ledger Explanation of Benefits Search.  SQL Ledger is a tool we chucked out about a million years ago.  The EOB part is also legacy, and was always wrong.  We hate the name.  You would be a hero if you got rid of it.  The file is also too darn big.

**Comment by MigDinny** [comment]
@aethelwulffe resolving the conflict + refactor the filename, is that what you are asking? If so, I can easily do it :)
If not, clarify me.

PS: i'm not a native english speaker, sometimes I don't understand your sentences and I need to read them twice to fully understand. It's normal if sometimes I don't get exactly what you meant :)

**Comment by aethelwulffe** [comment]
Either:
Resolve conflicts
OR
Refactor with new name.
Either one makes the conflict go away.
Do the one you prefer.  No need to change the file name if you don't want, but us old developers really hate seeing the "sl_" name everywhere.  Very old legacy references, and it is no longer a meaningful file name.

Terry might have a good name idea.  I think something like era_posting.php or era_posting.inc.php for the functions. (ERA stands for Electronic Remittance Advice, a file type that this stuff imports data from).

**Comment by MigDinny** [comment]
@aethelwulffe and what exactly would I have to do in that refactor? just change the name and its callings?

**Comment by aethelwulffe** [comment]
Yep...and delete the old file!

**Comment by MigDinny** [comment]
@aethelwulffe I can definitely do that, but, how will that solve the conflicts??

**Comment by aethelwulffe** [comment]
@MigDinny  Yes.  Creating a new file and deleting the one you worked on (along with changing the pointers where it is called elsewhere) resolves the issue.  Normally this is not the preferred method, except for the fact that giving this very important file a meaningful name would be additional value.

**Comment by MigDinny** [comment]
@aethelwulffe what name should I give it?

**Comment by aethelwulffe** [comment]
Try era_posting.php
@teryhill You got a better name?

**Comment by teryhill** [comment]
apply_payments    or posting    this will also be manual payments so era would only be half the description.

**Comment by MigDinny** [comment]
**apply_payments** or **era_posting** ? @teryhill  @aethelwulffe

**Comment by aethelwulffe** [comment]
Grandpa says apply_payments.php, and that is more accurate than my suggestion.

_**"Let it be written, let it be DONE!"**_

**Comment by MigDinny** [comment]
@aethelwulffe I've done it. Please check what I did because I'm not sure if that will work for you.

I've copied the master version of _sl_eob_search.php_ and I pasted it into _apply_payments.php_. I've made my changes and I've removed _sl_eob_search.php_

It says here that there are conflicts (see below) but I don't know how to solve them

image

**Comment by aethelwulffe** [comment]
Uh...that is because GIT, in all it's wisdom, recognized that you changed the name which didn't avoid the conflict.   Darn.
I will resolve the conflicts after we get another review approval.

File: interface/billing/apply_payments.php [hunk 7]
```diff
@@ -47,16 +47,15 @@
...
     /*  Addition of onsite portal patient notify of invoice and reformated invoice - sjpadgett 01/2017 */
     require_once("../../patient_portal/lib/portal_mail.inc");
     require_once("../../patient_portal/lib/appsql.class.php");
     
     function is_auth_portal( $pid = 0){
         if ($pData = sqlQuery("SELECT * FROM `patient_data` WHERE `pid` = ?", array($pid) )) {
-            if($pData['allow_patient_portal'] != "YES")
-                return false;
-                else return true;
+            if($pData['allow_patient_portal'] != "YES") return false;
```
Comment (teryhill) [inline interface/billing/apply_payments.php]: Would it be good to use the { } with the if statements. I am not sure how 7.1 and above are going to like this

**Comment by MigDinny** [comment]
@aethelwulffe i realized that git detected the filename being changed. it's far more intelligent than us, humans. 😆

File: interface/billing/apply_payments.php [hunk 8]
```diff
@@ -233,7 +232,7 @@ function upload_file_to_client_pdf($file_to_send) {
 
   // Print or download statements if requested.
   //
-if (($_POST['form_print'] || $_POST['form_download'] || $_POST['form_pdf']) || $_POST['form_portalnotify'] && $_POST['form_cb']) {
+  if (($_POST['form_print'] || $_POST['form_download'] || $_POST['form_pdf']) || $_POST['form_portalnotify'] && $_POST['form_cb']) {
```
Comment (teryhill) [inline interface/billing/apply_payments.php]: move the ) from the end of $_POST['form_pdf'} to the end of $_POST['form_portalnotify'] or is this in another PR

**Comment by MigDinny** [comment]
@teryhill done. I did it some commits ago but it appeared again probably because of the merges. I don't know. Check if it's okay now.

**Comment by robbyoconnor** [comment]
You're going to need to resolve conflicts

**Comment by aethelwulffe** [comment]
@MigDinny  Please have another go at getting this conflict resolved.  I can try if you like.  I would like to get this merged up.

**Comment by MigDinny** [comment]
@aethelwulffe my "resolve conflicts" button is disabled... I don't know how to resolve this... :/

**Comment by aethelwulffe** [comment]
Try restoring the original file (have it there at the same time as the new).  Commit that.  We can delete it afterwards.

**Comment by aethelwulffe** [comment]
https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/#removed-file-merge-conflicts

**Comment by MigDinny** [comment]
@aethelwulffe my git status does not say anything about conflicts

**Comment by MigDinny** [comment]
@aethelwulffe how can i restore the file? i'm completely lost

**Comment by MigDinny** [comment]
???????????????????????????????????????????????????????????????????????????

**Comment by MigDinny** [comment]
please don't call me a genius I dont know how i did that I've just merged master into this outdated branch

**Comment by aethelwulffe** [comment]
@MigDinny 
Please wrap the dangling IF statements (wrap in curly brackets) like Terry suggested, then let's get this moving again.

**Comment by MigDinny** [comment]
@aethelwulffe 9c633d1

**Comment by MigDinny** [comment]
@aethelwulffe 9c633d1

**Comment by MigDinny** [comment]
@aethelwulffe 9c633d1

**Comment by MigDinny** [comment]
@aethelwulffe 9c633d1

**Comment by MigDinny** [comment]
@aethelwulffe 9c633d1

**Comment by aethelwulffe** [comment]
@MigDinny 
This branch looks good, but we still have that darn conflict.
Anything you can do about it?  Maybe refactor this off of a fresh branch or anything?  This PR is very important!

**Comment by MigDinny** [comment]
@aethelwulffe Done

File: sites/default/sqlconf.php [hunk 9]
```diff
@@ -24,7 +24,7 @@
 //////////////////////////
 //////////////////////////
```
Comment (teryhill) [inline sites/default/sqlconf.php]: This file needs to be removed.

**Comment by MigDinny** [comment]
@teryhill Is it good now? I don't know if this last commit will actually remove the existing sqlconf or only remove it from this PR, i'm confused

**Comment by aethelwulffe** [comment]
We have two PR's related to sl_eob_search.  @teryhill this PR has the name change for that file.  You also have a later PR in on that.  Is it OK if we get this requested bug fix in (after resolving the sql_config thingie) before your insurance printing PR?

**Comment by teryhill** [comment]
@aethelwulffe  This needs to be closed and have @MigDinny  resubmit a clean PR.

**Comment by aethelwulffe** [comment]
@MigDinny I am going to give you until March 8 at 00:00 UTC to put in a corrected PR for this, and after that I am going to pull your code to my own branch and create a replacement PR.  I don't want the work lost, and we need to move on it now...So Sayeth Terry, so let it be done!

**Comment by MigDinny** [comment]
Hey @teryhill @aethelwulffe I'm a bit busy with school (really lots of stuff to study) and I was not checking my e-mail and with so many emails github ones got lost

I'll take a look into this

**Comment by MigDinny** [comment]
Conflicts solved @teryhill @aethelwulffe

**Comment by aethelwulffe** [comment]
Conflicts again.  This is a simple one-line change really, so I am squashing it to another PR.  It is good to have an "isset" check here just so weird stuff does not show up on the page...ever.

**Pull request closed without merging**